File: BFS/787. Cheapest Flights Within K Stops .cpp
/*

787. Cheapest Flights Within K Stops


There are n cities connected by m flights. Each fight starts from city u and arrives at v with a price w.

Now given all the cities and fights, together with starting city src and the destination dst, your task is to find the cheapest price from src to dst with up to k stops. If there is no such route, output -1.

Example 1:
Input: 
n = 3, edges = [[0,1,100],[1,2,100],[0,2,500]]
src = 0, dst = 2, k = 1
Output: 200
Explanation: 
The graph looks like this:


The cheapest price from city 0 to city 2 with at most 1 stop costs 200, as marked red in the picture.
Example 2:
Input: 
n = 3, edges = [[0,1,100],[1,2,100],[0,2,500]]
src = 0, dst = 2, k = 0
Output: 500
Explanation: 
The graph looks like this:


The cheapest price from city 0 to city 2 with at most 0 stop costs 500, as marked blue in the picture.
Note:

The number of nodes n will be in range [1, 100], with nodes labeled from 0 to n - 1.
The size of flights will be in range [0, n * (n - 1) / 2].
The format of each flight will be (src, dst, price).
The price of each flight will be in the range [1, 10000].
k is in the range of [0, n - 1].
There will not be any duplicated flights or self cycles.
*/


/*
graph store from i to j and time
BFS
everytime check the nearest point
每次像外延伸，用priority queue，每次先弄距离source cost最小的店, 
pq. [0] step to this point, [1] distance from source, [2] point

*/


class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        unordered_map<int, unordered_map<int,int>> graph;
        for(auto & flight: flights){
            graph[flight[0]][flight[1]] = flight[2];
        }
        vector<int>dist(n, numeric_limits<int>::max());
        dist[src] = 0;

        for(int a = 0; a <= k; ++a){
            vector<int>tmp = dist; //需要tmp 因为要只有dist 可能改变原来的
            for(int i = 0; i < n; ++i){
                if(dist[i] == numeric_limits<int>::max()) continue;
                for(auto& nxt: graph[i]){
                    tmp[nxt.first] = min(tmp[nxt.first], dist[i] + nxt.second);
                }
            }
            tmp.swap(dist);
        }
        return dist[dst] == numeric_limits<int>::max() ? -1: dist[dst];
    }
};


class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<vector<pair<int, int>>> adj(n);
        for (auto& e : flights) {
            adj[e[0]].push_back({e[1], e[2]});
        }
        vector<int> dist(n, numeric_limits<int>::max());
        queue<pair<int, int>> q;
        q.push({src, 0});
        int stops = 0;

        while (stops <= k && !q.empty()) {
            int sz = q.size();
            // Iterate on current level.
            while (sz--) {
                auto [node, distance] = q.front();
                q.pop();
                // Iterate over neighbors of popped node.
                for (auto& [neighbour, price] : adj[node]) {
                    if (price + distance >= dist[neighbour]) continue;
                    dist[neighbour] = price + distance;
                    q.push({neighbour, dist[neighbour]});
                }
            }
            stops++;
        }
        return dist[dst] == numeric_limits<int>::max() ? -1 : dist[dst];
    }
};


//TLE
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        vector<vector<pair<int, int>>> adj(n);
        for (auto e : flights) {
            adj[e[0]].push_back({e[1], e[2]});
        }
        vector<int> stops(n, numeric_limits<int>::max());
        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        // {dist_from_src_node, node, number_of_stops_from_src_node}
        pq.push({0, src, 0});

        while (!pq.empty()) {
            auto temp = pq.top();
            pq.pop();
            int dist = temp[0];
            int node = temp[1];
            int steps = temp[2];
            // We have already encountered a path with a lower cost and fewer stops,
            // or the number of stops exceeds the limit.
            if (steps > stops[node] || steps > k + 1) continue; 
            //if visited first, cost is the minimium 
            stops[node] = steps;
            if (node == dst) return dist;
            for (auto& [neighbor, price] : adj[node]) {
                pq.push({dist + price, neighbor, steps + 1});
            }
        }
        return -1;
    }
};

File: BFS/787. Cheapest Flights Within K Stops.py
"""
| Dijkstra's Algorithm

Need two dimensional visited. Or wrong. e.g. 

5
[[0,1,5],[1,2,5],[0,3,2],[3,1,2],[1,4,1],[4,2,1]]
0
2
2
"""

import heapq
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        dic = collections.defaultdict(list)
        for s, d, p in flights:
            dic[s].append([d, p])
        visited = collections.defaultdict(lambda: collections.defaultdict(int))
        h = [( 0,  src, 0)]  
        visited[src][-1] = 0
        while h:
            time, cur, count = heapq.heappop(h)
            if cur == dst:
                return time
            if count == K+1:
                continue 
            for nei, t in dic[cur]:
                if nei not in visited or count not in visited[nei] or visited[nei][count] > t + time:
                    visited[nei][count] = t+time
                    heapq.heappush(h,(time + t, nei, count+1))
        return -1


import heapq

class Solution:
    
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        
        # Build the adjacency matrix
        adj_matrix = [[0 for _ in range(n)] for _ in range(n)]
        for s, d, w in flights:
            adj_matrix[s][d] = w
            
        # Shortest distances array
        distances = [float("inf") for _ in range(n)]
        current_stops = [float("inf") for _ in range(n)]
        distances[src], current_stops[src] = 0, 0
        
        # Data is (cost, stops, node)
        minHeap = [(0, 0, src)]     
        
        while minHeap:
            
            cost, stops, node = heapq.heappop(minHeap)
            
            # If destination is reached, return the cost to get here
            if node == dst:
                return cost
            
            # If there are no more steps left, continue 
            if stops == K + 1:
                continue
             
            # Examine and relax all neighboring edges if possible 
            for nei in range(n):
                if adj_matrix[node][nei] > 0:
                    dU, dV, wUV = cost, distances[nei], adj_matrix[node][nei]
                    
                    # Better cost?
                    if dU + wUV < dV:
                        distances[nei] = dU + wUV
                        heapq.heappush(minHeap, (dU + wUV, stops + 1, nei))
                    elif stops < current_stops[nei]:
                        
                        #  Better steps?
                        current_stops[nei] = stops
                        heapq.heappush(minHeap, (dU + wUV, stops + 1, nei))
            
        return -1 if distances[dst] == float("inf") else distances[dst]

"""
DFS: 
Complexity:  O(V^2 K)


The number of recursive calls we can potentially make is O(VK).
In each recursive call, we iterate over a given node's neighbors. That takes time O(V) because we are using an adjacency matrix.

 Thus, the overall time complexity is O(V^2 K)

Space Complexity: O(VK + V + E)

O(VK ) is occupied by the memo dictionary and the rest by the adjacency matrix structure we build in the beginning.
"""

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, K: int) -> int:
        dic = collections.defaultdict(list)
        for s, d, p in flights:
            dic[s].append([d, p])
        visited = collections.defaultdict(lambda : collections.defaultdict(int))
        def dfs(cur, k):
            #print(cur, k)
            if k <0:
                return -1
            if cur == dst:
                return 0
            if cur in visited and k in visited[cur]:
                return visited[cur][k]
            res = -1
            for nxt, p in dic[cur]:
                ret = dfs(nxt, k-1)
                if ret != -1:
                    if res == -1:
                        res = p+ret
                    else:
                        res = min(res, p+ret)
            visited[cur][k] = res
            return res
        return dfs(src, K+1)